oak/server/oak_runtime.cc

```diff
@@ -35,31 +35,75 @@ asylo::Status OakRuntime::Initialize(const ApplicationConfiguration& config) {
   }
 
-  // TODO: Support creating multiple Nodes and Channels connecting them.
-
-  // Find first (only) Wasm node in the config (which must exist, because of
-  // the validity check above).
-  std::unique_ptr<OakNode> node = nullptr;
+  // Create all of the nodes.  The validity check above will ensure there
+  // is at most one of each pseudo-Node type.
+  std::string grpc_name;
+  std::string logging_name;
+  std::string storage_name;
   for (const auto& node_config : config.nodes()) {
     if (node_config.has_web_assembly_node()) {
-      node = OakNode::Create(node_config.web_assembly_node().module_bytes());
-      break;
+      LOG(INFO) << "Create Wasm node named " << node_config.node_name();
+      std::unique_ptr<OakNode> node =
+          OakNode::Create(node_config.node_name(), node_config.web_assembly_node().module_bytes());
+      if (node == nullptr) {
+        return asylo::Status(asylo::error::GoogleError::INVALID_ARGUMENT,
+                             "Failed to create Oak Node");
+      }
+      wasm_nodes_[node_config.node_name()] = std::move(node);
+    } else if (node_config.has_grpc_server_node()) {
+      grpc_name = node_config.node_name();
+      LOG(INFO) << "Create gRPC pseudo-Node named " << grpc_name;
+      grpc_node_ = OakGrpcNode::Create();
+    } else if (node_config.has_log_node()) {
+      logging_name = node_config.node_name();
+      LOG(INFO) << "Create logging pseudo-node named " << logging_name;
+      logging_node_ = absl::make_unique<LoggingNode>();
+    } else if (node_config.has_storage_node()) {
+      storage_name = node_config.node_name();
+      LOG(INFO) << "Created storage pseudo-Node named " << storage_name;
+      storage_node_ = absl::make_unique<StorageNode>(node_config.storage_node().address());
     }
   }
-  if (node == nullptr) {
-    return asylo::Status(asylo::error::GoogleError::INVALID_ARGUMENT, "Failed to create Oak Node");
-  }
 
-  // Setup the default channels for the node.
-  SetUpChannels(*node.get());
-
-  // TODO: This needs to be configurable
-  // Create the pseudo node required.
-  grpc_node_ = OakGrpcNode::Create();
-  SetUpGrpcChannels(*node.get());
+  // Now create channels.
+  for (const auto& channel_config : config.channels()) {
+    const std::string& src_node = channel_config.source_endpoint().node_name();
+    const std::string& src_port = channel_config.source_endpoint().port_name();
+    const std::string& dest_node = channel_config.destination_endpoint().node_name();
+    const std::string& dest_port = channel_config.destination_endpoint().port_name();
+
+    LOG(INFO) << "Create channel " << src_node << "." << src_port << " -> " << dest_node << "."
+              << dest_port;
+    auto channel = std::make_shared<MessageChannel>();
+    auto write_half = absl::make_unique<MessageChannelWriteHalf>(channel);
+    auto read_half = absl::make_unique<MessageChannelReadHalf>(channel);
+
+    // Wire up the source endpoint.
```

Comment:
This seems a very manual wiring, in general I would think all nodes should just be indexed in a map, and we just look up nodes by name and associate the channel to the port defined in the application configuration via `AddNamedChannel`.

Reply:
Yeah, I'm planning to homogenize nodes & pseudo-nodes later.  Added a TODO for now.